sparkletown/sparkle
# add I4A SSO via custom OAuth2 'Authorization Code Grant' flow

Add SSO via custom OAuth2 'Authorization Code Grant' flow.

fixes https://github.com/sparkletown/internal-sparkle-issues/issues/31

~~**Note:** This PR may need to be re-targeted onto `env/ohbm` depending on how it is finished off, as the I4A aspects of it are super specific.. Chunks of this would be useful for implementing a more generic custom OAuth provider though..~~ I think this is fine to go to `staging` based on how it's implemented now.

**Note:** When testing locally, you can't just use `npm start` as that won't map the 'rewrites' to the functions correctly. You need to use `npm run firebase:emulate-functions-hosting`.

## Reviewer Notes

If running locally to test, you need to:
  - Ensure your 'production' ENV values are mapped to staging
  - Comment out the `if (process.env.NODE_ENV === "development") {` check around `firebaseFunctions.useFunctionsEmulator("http://localhost:5001");`
  - Build the assets: `npm run build`
  - Run the function + hosting emulator: `npm run firebase:emulate-functions-hosting`
  - Access the login screen: http://localhost:5006/v/devaliashacksville
  - You should see the quicklogin icon:
    - ![image](https://user-images.githubusercontent.com/753891/122564423-abb7a100-d088-11eb-9c12-416e97b9187f.png)
  - Test credentials are available [here](https://doesitsparkle.slack.com/archives/C01ER4M61TL/p1624020034185200?thread_ts=1624019098.182500&cid=C01ER4M61TL)

If you don't do this, then you'll likely get a weird CORS error or similar, as per this slack thread: https://doesitsparkle.slack.com/archives/C01ER4M61TL/p1624019260183800

It seems as though the preview URL doesn't work for this either, with a similar CORS error. I don't believe that will happen in production, as it seems as though we get CORS errors from various thing on the platform in general when running on the preview deploys.

I have configured `/in/devaliashacksville/` to use the custom auth, which you can find in firebase in `authConfigs/devaliashacksville`

## TODO

- [x] Refactor `AUTH_ORIGIN` to be an array of whitelisted values on `authConfig` in firestore + read the 'redirect back to origin' from URL params + validate it against the whitelist: https://github.com/sparkletown/sparkle/pull/1523#discussion_r653470100 (https://github.com/sparkletown/sparkle/pull/1523/commits/a5147f88ee70065c69b17cabac44c9e058963088 + https://github.com/sparkletown/sparkle/pull/1523/commits/eb523e2c71d8dd19a4edbf50a2172f5bc5f48e78)
- [x] UI to actually redirect to I4A to initiate the login process
  - [x] Probably need a backend function to retrieve whatever details the frontend needs (eg. to check if it has some kind of auth config)
  - [x] Some `api/*` to call it
  - [x] Then integrate that in `src/pages/Account/Login.tsx` within the 'quick login' section to navigate to `/auth/connect/i4a?venueId=foo` (ideally returned from the aforementioned check)
- [ ] See TODO notes in this review: https://github.com/sparkletown/sparkle/pull/1523#pullrequestreview-685871766
  - [ ] Redirect back to platform and display 'not allowed, please register/buy a ticket' error: https://github.com/sparkletown/sparkle/pull/1523#discussion_r653233947
- [ ] Anything else?

### Potentially OOS?

- [ ] implement finding the sovereign venue efficiently/safely in backend functions, then use it to get the appropriate authConfig: https://github.com/sparkletown/sparkle/pull/1523#discussion_r653235698
  - Note: we could potentially get away without needing this for now so long as `authConfigs` are defined for all expected 'entrypoint' venues (which should already be the sovereign venues in most if not all cases anyway..)
- [ ] implement OAuth logout url correctly: https://github.com/sparkletown/sparkle/pull/1523#discussion_r653236835
